### ISSUE TITLE
@pnp/sp list items v3.4.0 works in every browser except Safari

### ISSUE THREAD

**Issue by kbeeveer46** [pnp/pnpjs#2315]
I just upgraded from v2 and everything is working correctly after refactoring the code to work for v3. I published the code to our QA SharePoint site catalog. When I view the QA site in Safari (I don't have Safari on my dev machine) it shows an error in place of where the web part is supposed to be. The developer console also displays several errors. 

I put the code that is causing the issue in the Step to Reproduce section.

Here is a screen shot of Chrome and Safari side-by-side on my M1 Macbook Pro. Safari is not showing the banner image. [https://i.imgur.com/xA9sWtG.png](https://i.imgur.com/xA9sWtG.png)

Here is a detailed screen shot of the errors (why is there a regular expression error in the dev console? I'm not using any)  [https://i.imgur.com/S7zwTr4.png](https://i.imgur.com/S7zwTr4.png)

### Category
- [ ] Enhancement
- [X] Bug
- [ ] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [ 3.4.0 ]

Please specify what version(s) of SharePoint you are targeting: [ 1.14.0 ]

### Expected / Desired Behavior / Question
The desired behavior is to query a web list by Id and select the items. 

### Observed Behavior
When the code runs it outputs an error message on the page and also several errors in the Safari developer console. 

### Steps to Reproduce
Here is the code that I am running. It's querying a list by id and getting the first item. It's getting the image property from the list item and returning it as a string to populate an img element

```
import { spfi, SPFx } from "@pnp/sp";
import "@pnp/sp/webs";
import "@pnp/sp/items";
import "@pnp/sp/lists";
import "@pnp/sp/files";
```

```
public async getBannerImage(): Promise<string>
    {
        let bannerImage = "";

        //const hub = Web(await this.hubSiteUrl.get());
        const hub = spfi(await this.hubSiteUrl.get()).using(SPFx(this._context));

        return new Promise<string>((resolve) =>
        {
            // "Home Page Banner" list

            hub.web.lists.getById("a866f8e4-eff0-4e9c-963b-f0b8613b6b94").items.select("Image_x0020_URL").top(1).orderBy("Created", true)().then((items: any[]) =>
            {
                if (items?.length)
                {
                    bannerImage = items[0]["Image_x0020_URL"].Url;

                    resolve(bannerImage);
                }

                resolve(bannerImage);

            }).catch(error =>
            {
                console.log("getBannerImage error: " + error);
                resolve(bannerImage);
            });
        });
    }
```

Here is the code which calls the method above

```
const Banner = (props: IBannerProps) =>
{

    // Props

    const {

        context

    } = props;

    // Services

    const homeService = new HomeService(context);

    // State

    const [bannerImage, setBannerImage] = useState("");

    useEffect(() =>
    {
        homeService.getBannerImage().then((bannerImage: string) =>
        {
            setBannerImage(bannerImage);
        });
    }, []);

    return <Image src={bannerImage} className="homeBannerImage" />;
};

export default Banner;
```

**Comment by koltyakov** [comment]
It actually doesn't seem to be something on PnPjs library side.

Based on an error on the screenshot looks that SPFx/infra related. A webpart bundle can't be loaded from CDN source.

**Comment by kbeeveer46** [comment]
> Based on an error on the screenshot looks that SPFx/infra related. A webpart bundle can't be loaded from CDN source.

The error message on the page (not dev console) seems to be a generic error that SharePoint uses. I've seen the same error pop up in completely different scenarios. When I revert the code back to v2 and re-upload the .sppkg file the error goes away. I'm 99% positive that something is wrong with v3 but I'm relatively new to SPFx so I can't say for certain. All I know is that it only happens when I'm using v3. If it was something wrong with SPFx why would it work in Edge and Chrome? (I forgot to test Firefox).

**Comment by juliemturner** [comment]
I'm sorry we would need a sample repository that we can use to recreate this issue as we cannot replicate it with just this small code snippet. I will say you're correct that you will often see that "generic" type error happening when your SPFx code throws an error that isn't caught before it can render, so I suspect you need to add some more error handling to your code to see where it's going wrong. 

Also, I'm going to offer a bit of a rewrite to you `getBannerImage` method, but again since I cannot see the complete code structure this may not be correct for your case...so obviously take it with a grain of salt.

```ts
public async getBannerImage(): Promise<string>{
  let bannerImage = "";
  try{
    //this line concerns me as I'm wondering what you're doing in this method
    const hubUrl = await this.hubSiteUrl.get(); 
    const hub = spfi(hubUrl).using(SPFx(this._context));
    // "Home Page Banner" list
    const items = await hub.web.lists.getById("a866f8e4-eff0-4e9c-963b-f0b8613b6b94").items.select("Image_x0020_URL").top(1).orderBy("Created", true)();
    if (items?.length){
      bannerImage = items[0]["Image_x0020_URL"].Url;
    }
  }catch(error) {
    console.log("getBannerImage error: " + error);
  }finally{
    return bannerImage;
  }
}
```

**Comment by kbeeveer46** [comment]
@juliemturner What's the best way to go about getting you a sample environment that replicates the issue? Creating a code sample on my end and sending that over shouldn't be an issue but since this code queries a hub site on the tenant at the company I work for how would I go about creating that environment for someone else to test against? Wouldn't that be something you would have to do on your end and then use my sample code above to see if the issue exists on your end? If I had my own SharePoint tenant/environment and had the sample code reach out to it wouldn't you run into permission issues? I'm not sure how to set up a test list that anyone could access using sample projects that replicates the issue.

In the meantime I'm going to see if I can find any sample projects on the PnP website that use list items to see if I can replicate the issue. 

Thanks for the code update, btw. I always forget "finally" exists. I'll try your code to see if the few modifications help at all. this.hubSiteUrl.get() is a property that calls an async method that uses PnPJs to get the hubsite URL. The method either retrieves it using PnP or cache. I'll try hardcoding the hub site URL and see if that changes anything.

**Comment by kbeeveer46** [comment]
@juliemturner I was able to get my project up and running on my Mac to be able to run the hosted workbench in Safari. The first thing I did was put console.log at the beginning of the getBannerImage method and after every line. To my surprise, it wasn't even hitting the first line of the method which tells me that it's not a problem with anything inside the method like the list items call.

I dug deeper into the regular express error (seen in my screen shot in the issue description) which is the first error that shows up when adding the web part to the page in Safari. Here is a screen shot of where it's coming from which looks to be part of the PnP SPQueryable class? Not sure if this is the root of the problem or a red herring. It appears that once this error happens the rest of the web part doesn't load. Which coincides with what you said about the generic SPFx error being shown before the render [https://i.imgur.com/56wwlvQ.png](https://i.imgur.com/56wwlvQ.png)

Was any of that code changed in v3 which may be causing the issue?

**Comment by juliemturner** [comment]
What I mean by a sample is a GitHub repository that has sample code in it that replicates the issue... so, you can simplify it if you want to certainly doesn't (and shouldn't) work against your tenant etc, but you need a codebase that's shareable that I can run in my environment to replicate the error.

We're seeing issues come up constantly with v3 and React Hooks and I'm sorry to say but it's almost always been an architecture issue and not a code issue... but certainly, there are exceptions. If I were you, I'd review the recommendations @bcameron1231 makes in the recently closed [2304](https://github.com/pnp/pnpjs/issues/2304#issuecomment-1149911244).

The problem with this issue is I think you're saying you're having problem only with the Safari browser on Mac and unfortunately neither @patrick-rodgers nor myself have a mac or have access to one.  So, I would say the only real hope we have is if someone with a mac can test your code and see if they can replicate the issue.

**Comment by kbeeveer46** [comment]
@juliemturner I appreciate you helping me out. I'm a little hesitant to say that it is an architecture issue because what I am using is just a single web part that displays an image on the page from a list using PnPJS. Similar to simple hello world web part you'd create when creating your first web part in a project. There's nothing to refactor and there's just a single context and spfi/SPFX object being used. I even converted it to a React Class component instead of a Hooks but I'm still seeing errors. Everything seems to work fine in Chrome, Edge, and FF (Windows and Mac). Just not Safari.

I guess the only thing left for me to do it put it into a sample project to see if anyone else can replicate it. If you are curious, here are the 2 files that I am using to create the error and a screen shot of the output. They are about as simple as you can get. They are essentially the same 2 files that are generated using yeoman for a new web part. 

Output: [https://i.imgur.com/T22Akk6.png](https://i.imgur.com/T22Akk6.png)

_BannerWebPart.ts (the initial file that is loaded when the web part is added to the page. This file loads the .tsx file below this and passes the context to it)_
```
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { Version } from '@microsoft/sp-core-library';
import { IPropertyPaneConfiguration } from '@microsoft/sp-property-pane';
import { BaseClientSideWebPart } from '@microsoft/sp-webpart-base';

import * as strings from 'BannerWebPartStrings';

import BannerNoHooks, { IBannerNoHooksProps } from './components/BannerNoHooks';

export default class BannerWebPart extends BaseClientSideWebPart<{}> {

    public onInit(): Promise<void>
    {
        return Promise.resolve();
    }

    public render(): void
    {
        const element: React.ReactElement<IBannerNoHooksProps> = React.createElement(
            BannerNoHooks,
            {
                context: this.context
            }
        );

        ReactDom.render(element, this.domElement);
    }

    protected onDispose(): void
    {
        ReactDom.unmountComponentAtNode(this.domElement);
    }

    protected get dataVersion(): Version
    {
        return Version.parse('1.0');
    }

    protected getPropertyPaneConfiguration(): IPropertyPaneConfiguration
    {
        return {
            pages: [
                {
                    header: {
                        description: strings.PropertyPaneDescription
                    },
                    groups: [
                        {
                            groupName: strings.BasicGroupName,
                            groupFields: []
                        }
                    ]
                }
            ]
        };
    }
}
```

_BannerNoHooks.tsx (the file that uses PnPJS and renders an image on the page)_
```
import { WebPartContext } from '@microsoft/sp-webpart-base';

// React
import * as React from 'react';

// PnP JS
import { spfi, SPFx } from "@pnp/sp";
import "@pnp/sp/webs";
import "@pnp/sp/items";
import "@pnp/sp/lists";

// Fluent UI
import { Image } from 'office-ui-fabric-react';

export interface IBannerNoHooksProps
{
    context: WebPartContext;
}

interface IBannerNoHooksState
{
    bannerImage: string;
}

export default class BannerNoHooks extends React.Component<IBannerNoHooksProps, IBannerNoHooksState, {}>
{
    constructor(props: IBannerNoHooksProps)
    {
        super(props);

        this.state = { bannerImage: "" };
    }

    public componentDidMount()
    {
        const hubUrl = "https://[REDACTED FOR PRIVACY].sharepoint.com/sites/ClientPortalHub";

        const hub = spfi(hubUrl).using(SPFx(this.props.context));

        hub.web.lists
            .getById("a866f8e4-eff0-4e9c-963b-f0b8613b6b94")
            .items
            .select("Image_x0020_URL")
            .top(1)
            .orderBy("Created", true)().then((items: any[]) =>
            {
                if (items?.length)
                    this.setState({ bannerImage: items[0]["Image_x0020_URL"].Url });

            });
    }

    public render(): React.ReactElement<IBannerNoHooksProps>
    {
        return <Image src={this.state.bannerImage} className="homeBannerImage" />;
    }
}
```

**Comment by kbeeveer46** [comment]
@juliemturner Here's a little more evidence that the "lookbehind" regex issue I'm seeing with PnPJS might be the root of the issue. The same regex was used in PnP Modern Search. The issue fixed in v4.2 in 2021. I'm 99.99% positive this is the same issue because in my workbench I see the same error this person is seeing. It's a different error than the generic error I was seeing on my QA site (I just happened to notice it was different a few minutes ago) [https://github.com/microsoft-search/pnp-modern-search/issues/685#issuecomment-816521062](https://github.com/microsoft-search/pnp-modern-search/issues/685#issuecomment-816521062)

**Comment by kbeeveer46** [comment]
@juliemturner I promise this is my last comment lol. @patrick-rodgers It looks like you committed the change that broke the PnPJS SP package in Safari when you made this commit to spqueryable.ts on April 22nd. The [lookbehind regex](https://caniuse.com/js-regexp-lookbehind) you added is not supported in Safari. Check out my previous comment in this thread to see how they fixed the same regex issue in PnP Modern Search [Commit 10d853f](https://github.com/pnp/pnpjs/commit/10d853ff9cd95a80da652f71b9be807fb68cd9cd#diff-bdc963fcde94659fffe052eb963737c9440c80ed8435a039055a7892e3d8a7bb)

I have never contributed to a GitHub project before and don't know how to push out the fixed code or else I would try and help. I appreciate everyone who helped me work through this today.

**Comment by juliemturner** [comment]
By architecture issue I mean architecture of the code... and as I said there are exceptions.

I'll investigate the issue with lookbehind regex not being supported and see if I can find one of the other pnp members who has a mac to try it out.

**Comment by juliemturner** [comment]
We've released a patch (3.4.1) that corrects this issue.  Thank you for reporting it.
